sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1409,48 +1409,7 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a Extract expression.
    */
   override def visitExtract(ctx: ExtractContext): Expression = withOrigin(ctx) {
-    ctx.field.getText.toUpperCase(Locale.ROOT) match {
-      case "MILLENNIUM" | "MILLENNIA" | "MIL" | "MILS" =>
-        Millennium(expression(ctx.source))
-      case "CENTURY" | "CENTURIES" | "C" | "CENT" =>
-        Century(expression(ctx.source))
-      case "DECADE" | "DECADES" | "DEC" | "DECS" =>
-        Decade(expression(ctx.source))
-      case "YEAR" | "Y" | "YEARS" | "YR" | "YRS" =>
-        Year(expression(ctx.source))
-      case "ISOYEAR" =>
-        IsoYear(expression(ctx.source))
-      case "QUARTER" | "QTR" =>
-        Quarter(expression(ctx.source))
-      case "MONTH" | "MON" | "MONS" | "MONTHS" =>
-        Month(expression(ctx.source))
-      case "WEEK" | "W" | "WEEKS" =>
-        WeekOfYear(expression(ctx.source))
-      case "DAY" | "D" | "DAYS" =>
-        DayOfMonth(expression(ctx.source))
-      case "DAYOFWEEK" =>
-        DayOfWeek(expression(ctx.source))
-      case "DOW" =>
-        Subtract(DayOfWeek(expression(ctx.source)), Literal(1))
-      case "ISODOW" =>
-        Add(WeekDay(expression(ctx.source)), Literal(1))
-      case "DOY" =>
-        DayOfYear(expression(ctx.source))
-      case "HOUR" | "H" | "HOURS" | "HR" | "HRS" =>
-        Hour(expression(ctx.source))
-      case "MINUTE" | "M" | "MIN" | "MINS" | "MINUTES" =>
-        Minute(expression(ctx.source))
-      case "SECOND" | "S" | "SEC" | "SECONDS" | "SECS" =>
-        Second(expression(ctx.source))
-      case "MILLISECONDS" | "MSEC" | "MSECS" | "MILLISECON" | "MSECONDS" | "MS" =>
-        Milliseconds(expression(ctx.source))
-      case "MICROSECONDS" | "USEC" | "USECS" | "USECONDS" | "MICROSECON" | "US" =>
-        Microseconds(expression(ctx.source))
-      case "EPOCH" =>
-        Epoch(expression(ctx.source))
-      case other =>
-        throw new ParseException(s"Literals of type '$other' are currently not supported.", ctx)
```

Comment:
The exception changed from `ParseException` to `AnalysisException`? Can we keep the current behaviour?

Reply:
`ParseException` requires either `ctx: ParserRuleContext` or `val start: Origin, val stop: Origin` that are not available to me at the point. And it is still `ParseException` in the output: https://github.com/apache/spark/pull/25410/files#diff-6f4edc80e2cc973e748705e85a6053b4R514

Reply:
... but I can pass `ctx: ParserRuleContext` to `DatePart`'s constructor. I am just not sure that is is good practice. WDYT?

Reply:
Instead of passing ctx into DataPart, can't we resolve `extractField` in `AstBuilder` then pass the resolved into DataPart?

Reply:
This will work for `extract` but how about the `date_part()` function itself? The same code should present for the function, shouldn't it?

Reply:
Ur, I see! sorry, but I need to be away from a keybord, so I'll recheck in hours.

Reply:
I've heard that its a little hard to read parser error messages in spark. So, I personally think we should throw ParseException with ctx for the `extract` case as much as possible. Then, how about writing it like this? https://github.com/apache/spark/commit/d793f4958f3a9adc69f7a1e04f7fc7a994d5e15a

Reply:
I agree it would be nice to point out the exact position of the error but I don't like this ad-hoc solution. There are a few similar places in `datetimeExpressions` and others where we could throw `ParseException` instead of just `AnalysisException`. I believe we need a common approach. Why do you think that propagation of the parsing context or positions look not good? We could propagate such parameters to all expression constructors.

Reply:
Ur, I don't think the approach is bad and I'm just looking for a better solution to keep the current error handling for that. If we already have the similar logic, can we follow that?

Reply:
@maropu @dongjoon-hyun If you think the commit https://github.com/apache/spark/commit/d793f4958f3a9adc69f7a1e04f7fc7a994d5e15a is necessary for merging the PR, I will apply the patch.